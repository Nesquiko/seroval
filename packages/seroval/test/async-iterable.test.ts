import { describe, it, expect } from 'vitest';
import {
  compileJSON,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  Feature,
  fromJSON,
  serializeAsync,
  toJSONAsync,
} from '../src';

describe('AsyncIterable', () => {
  describe('serializeAsync', () => {
    it('supports AsyncIterables', async () => {
      const example = {
        title: 'Hello World',
        async* [Symbol.asyncIterator](): AsyncIterator<number> {
          await Promise.resolve();
          yield 1;
          yield 2;
          yield 3;
        },
      };
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof example>(result);
      expect(back.title).toBe(example.title);
      expect(Symbol.asyncIterator in back).toBe(true);
      const iterator = back[Symbol.asyncIterator]();
      expect((await iterator.next()).value).toBe(1);
      expect((await iterator.next()).value).toBe(2);
      expect((await iterator.next()).value).toBe(3);
    });
  });
  describe('toJSONAsync', () => {
    it('supports AsyncIterables', async () => {
      const example = {
        title: 'Hello World',
        async* [Symbol.asyncIterator](): AsyncIterator<number> {
          await Promise.resolve();
          yield 1;
          yield 2;
          yield 3;
        },
      };
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof example>(result);
      expect(back.title).toBe(example.title);
      expect(Symbol.asyncIterator in back).toBe(true);
      const iterator = back[Symbol.asyncIterator]();
      expect((await iterator.next()).value).toBe(1);
      expect((await iterator.next()).value).toBe(2);
      expect((await iterator.next()).value).toBe(3);
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports AsyncIterables', async () => {
      const example = Promise.resolve({
        title: 'Hello World',
        async* [Symbol.asyncIterator]() {
          await Promise.resolve();
          yield 1;
          yield 2;
          yield 3;
        },
      });
      const result = await crossSerializeAsync(example);
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports AsyncIterables', async () => {
        const example = Promise.resolve({
          title: 'Hello World',
          async* [Symbol.asyncIterator]() {
            await Promise.resolve();
            yield 1;
            yield 2;
            yield 3;
          },
        });
        const result = await crossSerializeAsync(example, { scopeId: 'example' });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports AsyncIterables', async () => new Promise<void>((done) => {
      const example = Promise.resolve({
        title: 'Hello World',
        async* [Symbol.asyncIterator]() {
          await Promise.resolve();
          yield 1;
          yield 2;
          yield 3;
        },
      });
      crossSerializeStream(example, {
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          done();
        },
      });
    }));
    describe('scoped', () => {
      it('supports AsyncIterables', async () => new Promise<void>((done) => {
        const example = Promise.resolve({
          title: 'Hello World',
          async* [Symbol.asyncIterator]() {
            await Promise.resolve();
            yield 1;
            yield 2;
            yield 3;
          },
        });
        crossSerializeStream(example, {
          scopeId: 'example',
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            done();
          },
        });
      }));
    });
  });
  describe('compat', () => {
    it('should use function expressions instead of arrow functions.', async () => {
      const example = {
        async* [Symbol.asyncIterator](): unknown {
          await Promise.resolve();
          yield example;
        },
      };
      expect(await serializeAsync(example, {
        disabledFeatures: Feature.ArrowFunction,
      })).toMatchSnapshot();
    });
  });
  describe('compat#toJSONAsync', () => {
    it('should use function expression instead of arrow functions.', async () => {
      const example = {
        async* [Symbol.asyncIterator](): unknown {
          await Promise.resolve();
          yield example;
        },
      };
      const result = await toJSONAsync(example, {
        disabledFeatures: Feature.ArrowFunction,
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      expect(compileJSON(result)).toMatchSnapshot();
    });
  });
});
