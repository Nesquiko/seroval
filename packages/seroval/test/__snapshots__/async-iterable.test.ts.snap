// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AsyncIterable > compat > should use function expressions instead of arrow functions. 1`] = `"(function(h,j){return ({title:"Hello World",[h=Symbol.asyncIterator]:(function(s){return function(i,t){return (i=0,t={[h]:function(){return t},next:function(){return Promise.resolve().then(function(c,d){if(i>s.d)return{done:!0,value:void 0};c=i++,d=s.v[c];if(c===s.t)throw d;return{done:c===s.d,value:d}})}})}})({v:[1,2,3,void 0],t:-1,d:3})})})()"`;

exports[`AsyncIterable > compat#toJSONAsync > should use function expression instead of arrow functions. 1`] = `"{"t":{"t":10,"i":0,"p":{"k":["title",{"t":17,"i":1,"s":0}],"v":[{"t":1,"s":"Hello World"},{"t":45,"a":[{"t":44,"i":2,"s":0,"f":{"t":4,"i":1}},{"t":10,"i":3,"p":{"k":["v","t","d"],"v":[{"t":9,"i":4,"l":4,"a":[{"t":0,"s":1},{"t":0,"s":2},{"t":0,"s":3},{"t":2,"s":1}],"o":0},{"t":0,"s":-1},{"t":0,"s":3}],"s":3},"o":0}]}],"s":2},"o":0},"f":16379,"m":[1]}"`;

exports[`AsyncIterable > compat#toJSONAsync > should use function expression instead of arrow functions. 2`] = `"(function(h,j){return ({title:"Hello World",[h=Symbol.asyncIterator]:(function(s){return function(i,t){return (i=0,t={[h]:function(){return t},next:function(){return Promise.resolve().then(function(c,d){if(i>s.d)return{done:!0,value:void 0};c=i++,d=s.v[c];if(c===s.t)throw d;return{done:c===s.d,value:d}})}})}})({v:[1,2,3,void 0],t:-1,d:3})})})()"`;

exports[`AsyncIterable > crossSerializeAsync > scoped > supports AsyncIterables 1`] = `"($R=>$R[0]={title:"Hello World",[$R[1]=Symbol.asyncIterator]:($R[2]=s=>(i,t)=>(i=0,t={[$R[1]]:()=>t,next:()=>Promise.resolve().then((c,d)=>{if(i>s.d)return{done:!0,value:void 0};c=i++,d=s.v[c];if(c===s.t)throw d;return{done:c===s.d,value:d}})}))($R[3]={v:$R[4]=[1,2,3,void 0],t:-1,d:3})})($R["example"])"`;

exports[`AsyncIterable > crossSerializeAsync > supports AsyncIterables 1`] = `"$R[0]={title:"Hello World",[$R[1]=Symbol.asyncIterator]:($R[2]=s=>(i,t)=>(i=0,t={[$R[1]]:()=>t,next:()=>Promise.resolve().then((c,d)=>{if(i>s.d)return{done:!0,value:void 0};c=i++,d=s.v[c];if(c===s.t)throw d;return{done:c===s.d,value:d}})}))($R[3]={v:$R[4]=[1,2,3,void 0],t:-1,d:3})}"`;

exports[`AsyncIterable > crossSerializeStream > scoped > supports AsyncIterables 1`] = `"($R=>$R[0]={title:"Hello World",[$R[1]=Symbol.asyncIterator]:($R[2]=s=>(b,t)=>(b=s.tee(),s=b[0],b=b[1].getReader(),t={[$R[1]]:()=>t,next:()=>b.read().then(d=>{if(d.done)return{done:!0,value:void 0};d=d.value;if(d[0]===1)throw d[1];return{done:d[0]===2,value:d[1]}})}))($R[3]=($R[4]=(s,c)=>((s=new ReadableStream({start:x=>{c=x}})).c=c,s))())})($R["example"])"`;

exports[`AsyncIterable > crossSerializeStream > scoped > supports AsyncIterables 2`] = `"($R=>($R[6]=(s,d)=>{s.c.enqueue(d)})($R[3],$R[5]=[0,1]))($R["example"])"`;

exports[`AsyncIterable > crossSerializeStream > scoped > supports AsyncIterables 3`] = `"($R=>$R[6]($R[3],$R[7]=[0,2]))($R["example"])"`;

exports[`AsyncIterable > crossSerializeStream > scoped > supports AsyncIterables 4`] = `"($R=>$R[6]($R[3],$R[8]=[0,3]))($R["example"])"`;

exports[`AsyncIterable > crossSerializeStream > scoped > supports AsyncIterables 5`] = `"($R=>$R[6]($R[3],$R[9]=[2,void 0]))($R["example"])"`;

exports[`AsyncIterable > crossSerializeStream > scoped > supports AsyncIterables 6`] = `"($R=>($R[10]=s=>{s.c.close();delete s.c})($R[3]))($R["example"])"`;

exports[`AsyncIterable > crossSerializeStream > supports AsyncIterables 1`] = `"$R[0]={title:"Hello World",[$R[1]=Symbol.asyncIterator]:($R[2]=s=>(b,t)=>(b=s.tee(),s=b[0],b=b[1].getReader(),t={[$R[1]]:()=>t,next:()=>b.read().then(d=>{if(d.done)return{done:!0,value:void 0};d=d.value;if(d[0]===1)throw d[1];return{done:d[0]===2,value:d[1]}})}))($R[3]=($R[4]=(s,c)=>((s=new ReadableStream({start:x=>{c=x}})).c=c,s))())}"`;

exports[`AsyncIterable > crossSerializeStream > supports AsyncIterables 2`] = `"($R[6]=(s,d)=>{s.c.enqueue(d)})($R[3],$R[5]=[0,1])"`;

exports[`AsyncIterable > crossSerializeStream > supports AsyncIterables 3`] = `"$R[6]($R[3],$R[7]=[0,2])"`;

exports[`AsyncIterable > crossSerializeStream > supports AsyncIterables 4`] = `"$R[6]($R[3],$R[8]=[0,3])"`;

exports[`AsyncIterable > crossSerializeStream > supports AsyncIterables 5`] = `"$R[6]($R[3],$R[9]=[2,void 0])"`;

exports[`AsyncIterable > crossSerializeStream > supports AsyncIterables 6`] = `"($R[10]=s=>{s.c.close();delete s.c})($R[3])"`;

exports[`AsyncIterable > serializeAsync > supports AsyncIterables 1`] = `"((h,j)=>({title:"Hello World",[h=Symbol.asyncIterator]:(s=>(i,t)=>(i=0,t={[h]:()=>t,next:()=>Promise.resolve().then((c,d)=>{if(i>s.d)return{done:!0,value:void 0};c=i++,d=s.v[c];if(c===s.t)throw d;return{done:c===s.d,value:d}})}))({v:[1,2,3,void 0],t:-1,d:3})}))()"`;

exports[`AsyncIterable > toCrossJSONAsync > supports AsyncIterables 1`] = `"{"t":10,"i":0,"p":{"k":["title",{"t":17,"i":1,"s":0}],"v":[{"t":1,"s":"Hello World"},{"t":45,"a":[{"t":44,"i":2,"s":0,"f":{"t":4,"i":1}},{"t":10,"i":3,"p":{"k":["v","t","d"],"v":[{"t":9,"i":4,"l":4,"a":[{"t":0,"s":1},{"t":0,"s":2},{"t":0,"s":3},{"t":2,"s":1}],"o":0},{"t":0,"s":-1},{"t":0,"s":3}],"s":3},"o":0}]}],"s":2},"o":0}"`;

exports[`AsyncIterable > toCrossJSONStream > supports AsyncIterables 1`] = `"{"t":10,"i":0,"p":{"k":["title",{"t":17,"i":1,"s":0}],"v":[{"t":1,"s":"Hello World"},{"t":45,"a":[{"t":44,"i":2,"s":1,"f":{"t":4,"i":1}},{"t":31,"i":3,"f":{"t":41,"i":4,"s":5}}]}],"s":2},"o":0}"`;

exports[`AsyncIterable > toCrossJSONStream > supports AsyncIterables 2`] = `"{"t":32,"i":3,"a":[{"t":41,"i":6,"s":6},{"t":9,"i":5,"l":2,"a":[{"t":0,"s":0},{"t":0,"s":1}],"o":0}]}"`;

exports[`AsyncIterable > toCrossJSONStream > supports AsyncIterables 3`] = `"{"t":32,"i":3,"a":[{"t":4,"i":6},{"t":9,"i":7,"l":2,"a":[{"t":0,"s":0},{"t":0,"s":2}],"o":0}]}"`;

exports[`AsyncIterable > toCrossJSONStream > supports AsyncIterables 4`] = `"{"t":32,"i":3,"a":[{"t":4,"i":6},{"t":9,"i":8,"l":2,"a":[{"t":0,"s":0},{"t":0,"s":3}],"o":0}]}"`;

exports[`AsyncIterable > toCrossJSONStream > supports AsyncIterables 5`] = `"{"t":32,"i":3,"a":[{"t":4,"i":6},{"t":9,"i":9,"l":2,"a":[{"t":0,"s":2},{"t":2,"s":1}],"o":0}]}"`;

exports[`AsyncIterable > toCrossJSONStream > supports AsyncIterables 6`] = `"{"t":33,"i":3,"f":{"t":41,"i":10,"s":8}}"`;

exports[`AsyncIterable > toJSONAsync > supports AsyncIterables 1`] = `"{"t":{"t":10,"i":0,"p":{"k":["title",{"t":17,"i":1,"s":0}],"v":[{"t":1,"s":"Hello World"},{"t":45,"a":[{"t":44,"i":2,"s":0,"f":{"t":4,"i":1}},{"t":10,"i":3,"p":{"k":["v","t","d"],"v":[{"t":9,"i":4,"l":4,"a":[{"t":0,"s":1},{"t":0,"s":2},{"t":0,"s":3},{"t":2,"s":1}],"o":0},{"t":0,"s":-1},{"t":0,"s":3}],"s":3},"o":0}]}],"s":2},"o":0},"f":16383,"m":[1]}"`;
